cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/WorkflowStatsSlaHttpHandler.java

```diff
@@ -0,0 +1,173 @@
+
+    long startTime = TimeMathParser.parseTimeInSeconds(start);
+    long endTime = TimeMathParser.parseTimeInSeconds(end);
+
+    if (endTime < startTime || startTime < 0 || endTime < 0) {
+      responder.sendJson(HttpResponseStatus.BAD_REQUEST, "Wrong start or end time.");
+      return;
+    }
+
+    List<WorkflowDataset.WorkflowRunRecord> workflowRunRecords;
+
+    Id.Program programId = Id.Program.from(namespaceId, appId, ProgramType.WORKFLOW, workflowId);
+    try {
+      try {
+        workflowRunRecords = store.getWorkflowRuns(programId, startTime, endTime);
+
+      } catch (IllegalArgumentException exception) {
+        responder.sendJson(HttpResponseStatus.BAD_REQUEST, "Bad request");
+        return;
+      }
+    } catch (SecurityException exception) {
+      responder.sendJson(HttpResponseStatus.BAD_REQUEST, "You're not allowed to view this");
+      return;
+    }
+
+    int count = workflowRunRecords.size();
+
+    for (int i : percentiles) {
+      if (i > 100) {
+        responder.sendJson(HttpResponseStatus.BAD_REQUEST, "Wrong percentile values");
+        return;
+      }
+    }
+
+    if (count == 0) {
+      responder.sendJson(HttpResponseStatus.OK, "There were no completed runs for this workflow.");
+      return;
+    }
+
+    Collections.sort(workflowRunRecords, new Comparator<WorkflowDataset.WorkflowRunRecord>() {
+      @Override
+      public int compare(WorkflowDataset.WorkflowRunRecord o1, WorkflowDataset.WorkflowRunRecord o2) {
+        if (o1.getTimeTaken() > o2.getTimeTaken()) {
+          return 1;
+        } else if (o1.getTimeTaken() < o2.getTimeTaken()) {
+          return -1;
+        } else {
+          return 0;
+        }
+      }
+    });
+
+    Map<String, Long> workflowPercentileToTimeTaken = Maps.newHashMap();
+    Map<String, List<String>> slowestPercentileRuns = Maps.newHashMap();
+    for (int i : percentiles) {
+      List<String> percentileRun = new ArrayList();
+      int percentileStart = (i * count) / 100;
+      for (int j = percentileStart; j < count; j++) {
+        percentileRun.add(workflowRunRecords.get(j).getWorkflowRunId());
+      }
+      slowestPercentileRuns.put(Integer.toString(i), percentileRun);
+      workflowPercentileToTimeTaken.put(Integer.toString(i), workflowRunRecords.get(percentileStart).getTimeTaken());
+    }
+
+    Map<String, Object> response = Maps.newHashMap();
+    response.put("count", count);
+    response.put("percentile.completion", workflowPercentileToTimeTaken);
+    response.put("slowest.percentiles.runids", slowestPercentileRuns);
+
+    final Map<String, List<Long>> actionToRunRecord = Maps.newHashMap();
+    for (WorkflowDataset.WorkflowRunRecord workflowRunRecord: workflowRunRecords) {
+      for (WorkflowDataset.ActionRuns runs: workflowRunRecord.getActionRuns()) {
+        if (actionToRunRecord.get(runs.getName()) == null) {
+          List<Long> runList = Lists.newArrayList();
+          runList.add(runs.getTimeTaken());
```

Comment:
nit: seems like lines 149,150 are same as 153,154, so they can be pulled out of the if conditions